Mathlib/CategoryTheory/Sites/Preserves.lean

```diff
@@ -0,0 +1,169 @@
+  dsimp only [Equalizer.Presieve.Arrows.forkMap]
+  have h : Pi.lift (fun i ↦ F.map (c.inj i).op) =
+      F.map (Pi.lift (fun i ↦ (c.inj i).op)) ≫ piComparison F _ := by simp
+  rw [h, ← Category.assoc, ← Functor.map_comp]
+  have hh : Pi.lift (fun i ↦ (c.inj i).op) = (productIsProduct (op <| X ·)).lift c.op := by
+    simp [Pi.lift, productIsProduct]
+  rw [hh, ← desc_op_comp_opCoproductIsoProduct'_hom hc]
+  simp
+
+/--
+If `F` is a presheaf which `IsSheafFor` a presieve of arrows and the empty presieve, then it
+preserves the product corresponding to the presieve of arrows.
+-/
+noncomputable
+def preservesProductOfIsSheafFor (hF' : (ofArrows X c.inj).IsSheafFor F) :
+    PreservesLimit (Discrete.functor (fun x ↦ op (X x))) F := by
+  have : HasCoproduct X := ⟨⟨c, hc⟩⟩
+  refine @PreservesProduct.ofIsoComparison _ _ _ _ F _ (fun x ↦ op (X x)) _ _ ?_
+  rw [piComparison_fac (hc := hc)]
+  refine @IsIso.comp_isIso _ _ _ _ _ _ _ inferInstance ?_
+  rw [isIso_iff_bijective, Function.bijective_iff_existsUnique]
+  rw [Equalizer.Presieve.Arrows.sheaf_condition, Limits.Types.type_equalizer_iff_unique] at hF'
+  exact fun b ↦ hF' b (congr_fun (firstMap_eq_secondMap F hF hI c hd) b)
+
+/--
+If `F` preserves a particular product, then it `IsSheafFor` the corresponging presieve of arrows.
+-/
+theorem isSheafFor_of_preservesProduct [PreservesLimit (Discrete.functor (fun x ↦ op (X x))) F] :
+    (ofArrows X c.inj).IsSheafFor F := by
+  rw [Equalizer.Presieve.Arrows.sheaf_condition, Limits.Types.type_equalizer_iff_unique]
+  have : HasCoproduct X := ⟨⟨c, hc⟩⟩
+  have hi : IsIso (piComparison F (fun x ↦ op (X x))) := inferInstance
+  rw [piComparison_fac (hc := hc), isIso_iff_bijective, Function.bijective_iff_existsUnique] at hi
+  intro b _
+  obtain ⟨t, ht₁, ht₂⟩ := hi b
+  refine ⟨F.map ((opCoproductIsoProduct' hc (productIsProduct _)).inv) t, ht₁, fun y hy ↦ ?_⟩
+  apply_fun F.map ((opCoproductIsoProduct' hc (productIsProduct _)).hom) using injective_of_mono _
+  simp only [← FunctorToTypes.map_comp_apply, Iso.op, Category.assoc]
+  rw [ht₂ (F.map ((opCoproductIsoProduct' hc (productIsProduct _)).hom) y) (by simp [← hy])]
+  change (𝟙 (F.obj (∏ fun x ↦ op (X x)))) t = _
+  rw [← Functor.map_id]
+  refine congrFun ?_ t
+  congr
+  simp [Iso.eq_inv_comp, ← Category.assoc, ← op_comp, eq_comm, ← Iso.eq_comp_inv]
+
+/--
+A version of `isSheafFor_of_preservesProduct` for "the" coproduct instead of a general coproduct
+cone.
+-/
+theorem isSheafFor_of_preservesProduct' {Z : C} (π : (i : α) → X i ⟶ Z) [HasCoproduct X]
+    [IsIso (Sigma.desc π)] [PreservesLimit (Discrete.functor (fun x ↦ op (X x))) F]
+    [(ofArrows X π).hasPullbacks] : (ofArrows X π).IsSheafFor F :=
+  have : (ofArrows X (Cofan.mk Z π).inj).hasPullbacks :=
+    (inferInstance : (ofArrows X π).hasPullbacks)
+  have : IsIso (Sigma.desc (Cofan.mk Z π).inj) := (inferInstance : IsIso (Sigma.desc π))
+  isSheafFor_of_preservesProduct F (Cofan.mk Z π) (Cofan.mk Z π).isColimitOfIsIsoSigmaDesc
+
```

Comment:
Do you plan to use this lemma in future PR? If not, we may remove it.
```suggestion
```

Reply:
I use it in #7949

Reply:
I'll probably need to refactor that PR a bit anyway once this is merged, and maybe I can just use the un-primed version of this theorem directly. Let's remove it for now and I'll add it if I really need it.